src/main/io/rcdevice.c

```diff
@@ -0,0 +1,668 @@
+}
+
+static bool runcamDeviceDecodeSettings(sbuf_t *buf, runcamDeviceSetting_t outSettingList[RCDEVICE_PROTOCOL_MAX_MENUITEM_PER_PAGE])
+{
+    if (outSettingList == NULL) {
+        return false;
+    }
+
+    runcamDeviceSettingParseStep_e parseStep = RCDP_SETTING_PARSE_WAITING_ID;
+    memset(outSettingList, 0, RCDEVICE_PROTOCOL_MAX_MENUITEM_PER_PAGE * sizeof(runcamDeviceSetting_t));
+    runcamDeviceSetting_t *settingIterator = outSettingList;
+    int i = 0;
+    while (sbufBytesRemaining(buf) > 0) {
+        if (i >= RCDEVICE_PROTOCOL_MAX_MENUITEM_PER_PAGE) {
+            break;
+        }
+
+        switch (parseStep) {
+        case RCDP_SETTING_PARSE_WAITING_ID: {
+            uint8_t c = sbufReadU8(buf);
+            settingIterator->id = c;
+            parseStep = RCDP_SETTING_PARSE_WAITING_NAME;
+        } 
+            break;
+        case RCDP_SETTING_PARSE_WAITING_NAME: {
+            
+            const char *str = (const char *)sbufConstPtr(buf);
+            uint8_t nameLen = strlen(str) + 1;
+            memset(settingIterator->name, 0, RCDEVICE_PROTOCOL_MAX_SETTING_NAME_LENGTH);
+            strncpy(settingIterator->name, str, RCDEVICE_PROTOCOL_MAX_SETTING_NAME_LENGTH);
+            sbufAdvance(buf, nameLen);
+            
+            parseStep = RCDP_SETTING_PARSE_WAITING_VALUE;
+        } 
+            break;
+        case RCDP_SETTING_PARSE_WAITING_VALUE: {
+            const char *str = (const char *)sbufConstPtr(buf);
+            uint8_t valueLen = strlen(str) + 1;
+            memset(settingIterator->value, 0, RCDEVICE_PROTOCOL_MAX_SETTING_VALUE_LENGTH);
+            strcpy(settingIterator->value, str);
+            sbufAdvance(buf, valueLen);
+            parseStep = RCDP_SETTING_PARSE_WAITING_ID;
+
+            settingIterator++;
+            i++;
+        } 
+            break;
+        }
+    }
+
+    if (RCDP_SETTING_PARSE_WAITING_ID != parseStep) {
+        return false;
+    }
+
+    return true;
+}
+
+static bool runcamDeviceGetResponseWithMultipleChunk(runcamDevice_t *device, uint8_t command, uint8_t settingID, uint8_t *responseData, uint16_t *responseDatalen)
+{
+    if (responseData == NULL || responseDatalen == NULL) {
+        return false;
+    }
+
+    // fill parameters buf
+    uint8_t paramsBuf[2];
+    uint8_t chunkIndex = 0;
+    paramsBuf[0] = settingID; // parent setting id
+    paramsBuf[1] = chunkIndex; // chunk index
+
+    uint8_t outputBufLen = RCDEVICE_PROTOCOL_MAX_PACKET_SIZE;
+    uint8_t outputBuf[RCDEVICE_PROTOCOL_MAX_PACKET_SIZE];
+    bool result = runcamDeviceSendRequestAndWaitingResp(device, command, paramsBuf, sizeof(paramsBuf), outputBuf, &outputBufLen);
+    if (!result) {
+        return false;
+    }
+
+    uint8_t remainingChunk = outputBuf[1];
+    // Every response chunk count must less than or equal to RCDEVICE_PROTOCOL_MAX_CHUNK_PER_RESPONSE
+    if ((remainingChunk + 1) > RCDEVICE_PROTOCOL_MAX_CHUNK_PER_RESPONSE) {
```

Comment:
`remainingChunk >= RCDEVICE_PROTOCOL_MAX_CHUNK_PER_RESPONSE` ?

Reply:
The condition (remainingChunk + 1) > RCDEVICE_PROTOCOL_MAX_CHUNK_PER_RESPONSE(12) is right, e.g there is a response, and it has splited as 12 chunks, so the remaining chunk filed on the every chunk looks like:
chunk0: remaining chunk count(11)
chunk1: remaining chunk count(10)
chunk2: remaining chunk count(9)
chunk3: remaining chunk count(8)
....
chunk11: remaining chunk count(0)

Reply:
Results of both are equivalent ( ` +1 >` / `>=`). It's just that `<` test is used in C to check if value is within bounds and `>=` is its inverse (checking for out-of-bounds)...

Reply:
You are right! Thank you!